Mathlib/Algebra/Category/ModuleCat/TensorHomAdjunction.lean

```diff
@@ -0,0 +1,213 @@
+But currently mathlib4 does not know `Y⊗[R] X` is an right `S`-module as well.
+
+2. Why not use the [`Tower` file](Mathlib/LinearAlgebra/TensorProduct/Tower.lean`)
+
+In our setting `X` is an `(R, S)`-bimodule and `Y` an `R`-module and `Z` an `S`-module
+so to use the `Tower` file, we need `S` to be an `R`-algebra which is a luxury we do not have.
+But note that in `Tower` file, `curry` and `uncurry` are both tri-linear maps. So `Tower` file
+allows interplay of 3 rings which is not allowed in this file.
+
+-/
+
+universe u u' v v' v''
+
+open TensorProduct CategoryTheory
+
+variable (R : Type u) (S : Type u') (X : Type v)
+variable [CommRing R] [Ring S]
+variable [AddCommGroup X] [Module Rᵐᵒᵖ X] [Module S X] [SMulCommClass S Rᵐᵒᵖ X]
+
+open MulOpposite
+
+-- TODO: remove these local instances when TensorProduct is generalized over non-commutative rings.
+
+/-- The `R`-module structure on a `Rᵐᵒᵖ`-module `X` for a commutative ring `R`. -/
+def Module.ofMop : Module R X :=
+  Module.compHom X ((RingHom.id R).toOpposite fun _ _ => mul_comm _ _)
+
+attribute [local instance] Module.ofMop
+
+variable {Y : Type v'} [AddCommGroup Y] [Module R Y]
+
+/-- A version of `TensorProduct.leftModule` with `Rᵐᵒᵖ` in place of `R`. -/
+noncomputable def Module.tensorLeftOfMop : Module S (X ⊗[R] Y) :=
+  have : SMulCommClass R S X :=
+    ⟨fun r s x ↦ show (op r) • s • x = s • (op r) • x by rw [← smul_comm]⟩
+  inferInstance
+
+attribute [local instance] Module.tensorLeftOfMop
+
+@[simp] lemma tensorModSMul_smul_tmul
+  (s : S) (x : X) (y : Y) : s • (x ⊗ₜ[R] y) = (s • x) ⊗ₜ[R] y := rfl
+
+variable {Z : Type v''} [AddCommGroup Z] [Module S Z]
+/--
+Let `R` be a commutative ring and `S` a ring.
+Given an `(S,R)`-bimodule `X` and a left `S`-module `Y`, the set of
+`S`-linear maps from `X` to `Y` has a left `R`-module structure given by:
+```
+l : X →ₗ[S] Y
+r : R
+x : X
+-------------
+(r • l) x = l (op r • x)
+```
+-/
+def Module.linearMapLeft : Module R (X →ₗ[S] Z) where
+  smul r l :=
+  { toFun := fun x => l (op r • x)
+    map_add' := fun x y => by dsimp; rw [smul_add, map_add]
+    map_smul' := fun s x => by dsimp; rw [← smul_comm, l.map_smul] }
+  one_smul l := LinearMap.ext fun x => show l _ = _ by rw [op_one, one_smul]
+  mul_smul r₁ r₂ l := LinearMap.ext fun x => show l _ = l _ by rw [op_mul, mul_smul]
+  smul_zero r := rfl
+  smul_add r l₁ l₂ := LinearMap.ext fun x => show (l₁ + _) _ = _ by
+    rw [LinearMap.add_apply, LinearMap.add_apply]; rfl
+  add_smul r₁ r₂ l := LinearMap.ext fun x => show l _ = l _ + l _ by
+    rw [op_add, add_smul, map_add]
+  zero_smul l := LinearMap.ext fun x => show l _ = 0 by rw [op_zero, zero_smul, map_zero]
+
+attribute [local instance] Module.linearMapLeft
+
+variable {R S X}
+/--
+Let `R` be a commutative ring and `S` a ring.
+Given an `(S,R)`-bimodule `X`, a left `R`-module `Y` and a left `S`-module `Z`,
+any `S`-linear map `X ⊗[R] Y ⟶ Z` can by curried into a bilinear map `Y ⟶ X ⟶ Z` that
```

Comment:
In my opinion this is a rather weird meaning of "bilinear"

Reply:
Indeed when I modify the docstring I didn't realize it doesn't make perfect sense to say it's `R`-linear in the first argument, as `Z` has no `R`-module structure.